internal/serialization/serialization_improvements_test.go

```diff
@@ -119,11 +120,31 @@ func TestSerializationImprovements(t *testing.T) {
 				name:   "JavaOffsetDateTime",
 				target: time.Date(2021, 2, 10, 1, 2, 3, 4, time.FixedZone("", -3*60*60)),
 			},
+			{
+				input:  []interface{}{"foo", int64(22)},
+				name:   "JavaArrayList",
+				target: []interface{}{"foo", int64(22)},
+			},
 			{
 				input:  big.NewInt(-10_000_000),
 				name:   "JavaBigInteger",
 				target: big.NewInt(-10_000_000),
 			},
```

Comment:
Can we add this to here as well?
```go
{
	input:  []interface{}{},
	name:   "JavaArrayList-Empty",
	target: []interface{}{},
},
```

 As far as I see this fails due to a bug in array list serializer. When the slice's length is 0 we should not write -1 to the length. You can check java's ArrayListStreamSerializer. We were using NULL_ARRAY_LENGTH in serialization before 2019. It is changed in this pr: https://github.com/hazelcast/hazelcast/pull/15371/files
You might have checked python and take implementation from there. Python as well as Nodejs serialization code seems a bit out of date regarding these changes. Maybe we should sync them with server soon.

Reply:
You're right. Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/724/commits/c331dcb2b8fff769657f40b733f9b2111336676b